### ISSUE TITLE
3.0 branch broken following merge of #21971

### ISSUE THREAD

**Issue by tom-cosgrove-arm** [openssl/openssl#22114]
The merge to 3.0 has broken the 3.0 branch

```
crypto/engine/eng_list.c:85:49: error: too few arguments provided to function-like macro invocation
            CRYPTO_DOWN_REF(&e->struct_ref, &ref);
                                                ^
include/internal/refcount.h:162:10: note: macro 'CRYPTO_DOWN_REF' defined here
# define CRYPTO_DOWN_REF(val, ret, lock) CRYPTO_atomic_add(val, -1, ret, lock)
         ^
```

as the declaration of `ref` was not brought in

```
diff --git a/crypto/engine/eng_list.c b/crypto/engine/eng_list.c
index 04c73c7628..d3bd0c0dc2 100644
--- a/crypto/engine/eng_list.c
+++ b/crypto/engine/eng_list.c
@@ -78,12 +78,16 @@ static int engine_list_add(ENGINE *e)
             ERR_raise(ERR_LIB_ENGINE, ENGINE_R_INTERNAL_LIST_ERROR);
             return 0;
         }
-        engine_list_head = e;
-        e->prev = NULL;
         /*
          * The first time the list allocates, we should register the cleanup.
          */
-        engine_cleanup_add_last(engine_list_cleanup);
+        if (!engine_cleanup_add_last(engine_list_cleanup)) {
+            CRYPTO_DOWN_REF(&e->struct_ref, &ref);
+            ERR_raise(ERR_LIB_ENGINE, ENGINE_R_INTERNAL_LIST_ERROR);
+            return 0;
+        }
+        engine_list_head = e;
+        e->prev = NULL;
     } else {
         /* We are adding to the tail of an existing list. */
         if ((engine_list_tail == NULL) || (engine_list_tail->next != NULL)) {
```

**Comment by tom-cosgrove-arm** [comment]
I consider this urgent - I can create a fix, but internal approval process means it would take a while to get in, whereas I can review much faster

**Comment by bernd-edlinger** [comment]
Fixed by #22115
